src/SignalR/clients/cpp/src/signalrclient/connection_impl.cpp

```diff
@@ -84,94 +84,139 @@ namespace signalr
-                        return pplx::task_from_result();
-                    });
-            }, m_disconnect_cts.get_token())
-            .then([start_tce, weak_connection](pplx::task<void> previous_task)
+            // TODO: use transfer format
+
+            return connection->start_transport(url)
+                .then([weak_connection, start_tce](std::shared_ptr<transport> transport)
             {
                 auto connection = weak_connection.lock();
                 if (!connection)
                 {
                     return pplx::task_from_exception<void>(_XPLATSTR("connection no longer exists"));
                 }
-                try
-                {
-                    previous_task.get();
-                    if (!connection->change_state(connection_state::connecting, connection_state::connected))
-                    {
-                        connection->m_logger.log(trace_level::errors,
-                            utility::string_t(_XPLATSTR("internal error - transition from an unexpected state. expected state: connecting, actual state: "))
-                            .append(translate_connection_state(connection->get_connection_state())));
+                connection->m_transport = transport;
 
-                        _ASSERTE(false);
-                    }
-
-                    connection->m_start_completed_event.set();
-                    start_tce.set();
-                }
-                catch (const std::exception &e)
+                if (!connection->change_state(connection_state::connecting, connection_state::connected))
                 {
-                    auto task_canceled_exception = dynamic_cast<const pplx::task_canceled *>(&e);
-                    if (task_canceled_exception)
-                    {
-                        connection->m_logger.log(trace_level::info,
-                            _XPLATSTR("starting the connection has been canceled."));
-                    }
-                    else
-                    {
-                        connection->m_logger.log(trace_level::errors,
-                            utility::string_t(_XPLATSTR("connection could not be started due to: "))
-                            .append(utility::conversions::to_string_t(e.what())));
-                    }
+                    connection->m_logger.log(trace_level::errors,
+                        utility::string_t(_XPLATSTR("internal error - transition from an unexpected state. expected state: connecting, actual state: "))
+                        .append(translate_connection_state(connection->get_connection_state())));
 
-                    connection->m_transport = nullptr;
-                    connection->change_state(connection_state::disconnected);
-                    connection->m_start_completed_event.set();
-                    start_tce.set_exception(std::current_exception());
+                    _ASSERTE(false);
                 }
 
                 return pplx::task_from_result();
```

Comment:
Why return a completed task here only to immediately continue into the following then block? Couldn't you just put the following lines directly here instead and save yourself all the then block boiler plate to reacquire the connection from the weak reference and so forth below?

```
connection->m_start_completed_event.set();
start_tce.set();
```

Reply:
Nope, if you look closely the `.then` is actually for a bigger scope and will handle cleanup for failures from earlier parts of the negotiate logic.